src/controllers/plugins/api/IOWrapper.cpp

```diff
@@ -0,0 +1,342 @@
+        if (major != 'r' && major != 'w' && major != 'a')
+        {
+            this->error = "Invalid mode, one matching [rwa][+]?b?. "
+                          "Parsing failed at 1st character";
+            return;
+        }
+        this->major = major.toLatin1();
+        if (smode.length() > 1)
+        {
+            auto plusOrB = smode.at(1);
+            if (plusOrB == '+')
+            {
+                this->update = true;
+            }
+            else if (plusOrB == 'b')
+            {
+                this->binary = true;
+            }
+            else
+            {
+                this->error = "Invalid mode, one matching [rwa][+]?b?. "
+                              "Parsing failed at 2nd character";
+                return;
+            }
+        }
+        if (smode.length() > 2)
+        {
+            auto maybeB = smode.at(2);
+            if (maybeB == 'b')
+            {
+                this->binary = true;
+            }
+            else
+            {
+                this->error = "Invalid mode, one matching [rwa][+]?b?. "
+                              "Parsing failed at 3rd character";
+                return;
+            }
+        }
+    }
+
+    QString toString() const
+    {
+        assert(this->major == 'r' || this->major == 'w' || this->major == 'a');
+        QString out;
+        out += this->major;
+        if (this->update)
+        {
+            out += '+';
+        }
+        if (this->binary)
+        {
+            out += 'b';
+        }
+        return out;
+    }
+};
+
+// NOLINTBEGIN(*vararg)
+int io_open(lua_State *L)
+{
+    auto *pl = getApp()->getPlugins()->getPluginByStatePtr(L);
+    if (pl == nullptr)
+    {
+        luaL_error(L, "internal error: no plugin");
+        return 0;
+    }
+    LuaFileMode mode;
+    if (lua_gettop(L) == 2)
+    {
+        // we have a mode
+        QString smode;
+        if (!lua::pop(L, &smode))
+        {
+            return luaL_error(
+                L,
+                "io.open mode (2nd argument) must be a string or not present");
+        }
+        mode = LuaFileMode(smode);
+        if (!mode.error.isEmpty())
+        {
+            return luaL_error(L, mode.error.toStdString().c_str());
+        }
+    }
+    QString filename;
+    if (!lua::pop(L, &filename))
+    {
+        return luaL_error(L,
+                          "io.open filename (1st argument) must be a string");
+    }
+    QFileInfo file(pl->loadDirectory().filePath(filename));
+    auto abs = file.absoluteFilePath();
+    qCDebug(chatterinoLua) << "[" << pl->id << ":" << pl->meta.name
+                           << "] Plugin is opening file at " << abs
+                           << " with mode " << mode.toString();
+    bool ok = pl->hasFSPermissionFor(
+        mode.update || mode.major == 'w' || mode.major == 'a', abs);
+    if (!ok)
+    {
+        return luaL_error(
+            L, "Plugin does not have permissions to access given file.");
```

Comment:
These shouldn't be runtime errors but `fail` and an error message (see [docs](https://www.lua.org/manual/5.4/manual.html#6.8)).